src/programs.rkt

```diff
@@ -205,15 +201,17 @@
         (define fn (operator-info op 'nonffi))
         (define res (apply fn args))
         (define rtype (operator-info op 'otype))
-        (and ((value-of rtype) res)
-             (exact-value? rtype res)
-             (value->code rtype res)))
+        ;; TODO: splitting real into binary64 and binary32 problems
+        (define rtype* (if (set-member? '(binary64 binary32) rtype) 'real rtype))
+        (and ((value-of rtype*) res)
+             (exact-value? rtype* res)
+             (value->code rtype* res)))
```

Comment:
IIRC, `value-of` and `exact-value?`, and `value->code`, are all called here. So we could remove this `rtype*` thing and push it into those functions.

Also, uhh, I think this ugly code is only here to support constant-folding on complex numbers, and I'm not sure I care much. Trello card added.